application/src/main/java/org/thingsboard/server/service/install/DefaultDatabaseSchemaSettingsService.java

```diff
@@ -0,0 +1,146 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.thingsboard.server.service.install;
+
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.boot.info.BuildProperties;
+import org.springframework.context.annotation.Profile;
+import org.springframework.dao.EmptyResultDataAccessException;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.stereotype.Service;
+import org.thingsboard.server.service.install.update.DefaultDataUpdateService;
+
+import java.util.List;
+
+@Service
+@Profile("install")
+@Slf4j
+@RequiredArgsConstructor
+public class DefaultDatabaseSchemaSettingsService implements DatabaseSchemaSettingsService {
+
+    private static final String CURRENT_PRODUCT = "CE";
+    private static final List<String> SUPPORTED_VERSIONS_FOR_UPGRADE = List.of("3.8.0", "3.8.1");
+
+    private final BuildProperties buildProperties;
+    private final JdbcTemplate jdbcTemplate;
+
+    private String packageSchemaVersion;
+    private String schemaVersionFromDb;
+
+    @Override
+    public void validateSchemaSettings() {
+        //TODO: remove after release (3.9.0)
+        createProductIfNotExists();
+
+        String dbSchemaVersion = getDbSchemaVersion();
+
+        if (DefaultDataUpdateService.getEnv("SKIP_SCHEMA_VERSION_CHECK", false)) {
+            log.info("Skipped DB schema version check due to SKIP_SCHEMA_VERSION_CHECK set to 'true'.");
+            return;
+        }
+
+        String product = getProductFromDb();
+        if (!CURRENT_PRODUCT.equals(product)) {
+            onSchemaSettingsError(String.format("Upgrade failed: can't upgrade ThingsBoard %s database using ThingsBoard %s.", product, CURRENT_PRODUCT));
+        }
+
+        if (dbSchemaVersion.equals(getPackageSchemaVersion())) {
+            onSchemaSettingsError("Upgrade failed: database already upgraded to current version. You can set SKIP_SCHEMA_VERSION_CHECK to 'true' if force re-upgrade needed.");
+        }
+
+        if (!SUPPORTED_VERSIONS_FOR_UPGRADE.contains(dbSchemaVersion)) {
+            onSchemaSettingsError(String.format("Upgrade failed: database version '%s' is not supported for upgrade. Supported versions are: %s.",
+                    dbSchemaVersion, SUPPORTED_VERSIONS_FOR_UPGRADE
+            ));
+        }
+    }
+
+    @Deprecated(forRemoval = true, since = "3.9.0")
+    private void createProductIfNotExists() {
+        boolean isCommunityEdition;
+        try {
+            jdbcTemplate.queryForObject("SELECT 1 FROM information_schema.tables WHERE table_name = 'integration'", Integer.class);
+            isCommunityEdition = false;
+        } catch (EmptyResultDataAccessException e) {
```

Comment:
I guess that there will be an error log for this. Use queryForList or smth like that to avoid this.